x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java

```diff
@@ -645,27 +645,50 @@ private static void checkFilterMatchConditions(LogicalPlan plan, Set<Failure> fa
     private static void checkFullTextQueryFunctions(LogicalPlan plan, Set<Failure> failures) {
         if (plan instanceof Filter f) {
             Expression condition = f.condition();
-            if (condition instanceof FullTextFunction ftf) {
-                // Similar to cases present in org.elasticsearch.xpack.esql.optimizer.rules.PushDownAndCombineFilters -
-                // we can't check if it can be pushed down as we don't have yet information about the fields present in the
-                // StringQueryPredicate
-                plan.forEachDown(LogicalPlan.class, lp -> {
-                    if ((lp instanceof Filter || lp instanceof OrderBy || lp instanceof EsRelation) == false) {
-                        failures.add(
-                            fail(
-                                plan,
-                                "[{}] function cannot be used after {}",
-                                ftf.functionName(),
-                                lp.sourceText().split(" ")[0].toUpperCase(Locale.ROOT)
-                            )
-                        );
-                    }
-                });
+            Holder<FullTextFunction> functionHolder = new Holder<>();
+            condition.forEachDown(FullTextFunction.class, functionHolder::set);
+            FullTextFunction ftf = functionHolder.get();
+            if (ftf != null) {
+                checkCommandUsageAfterFullTextFunction(ftf, plan, failures);
+                checkConditionCanBePushedDown(ftf, condition, failures);
             }
         } else {
             plan.forEachExpression(FullTextFunction.class, ftf -> {
                 failures.add(fail(ftf, "[{}] function is only supported in WHERE commands", ftf.functionName()));
             });
         }
     }
+
+    private static void checkCommandUsageAfterFullTextFunction(FullTextFunction ftf, LogicalPlan plan, Set<Failure> failures) {
+        // Similar to cases present in org.elasticsearch.xpack.esql.optimizer.rules.PushDownAndCombineFilters -
+        // we can't check if it can be pushed down if we don't have information about the fields present in the query.
+        plan.forEachDown(LogicalPlan.class, lp -> {
+            if ((lp instanceof Filter || lp instanceof OrderBy || lp instanceof EsRelation || ftf.hasFieldsInformation()) == false) {
```

Comment:
I found the checking to `ftf.hasFieldsInformation()` a bit confusing, I wonder if it means these restrictions apply to `QueryStringFunction` only? Does it mean `MatchFunction` can be after any processing commands? 

Some comments will be helpful, or if this applies to `QueryStringFunction` only and `hasFieldsInformation()` is only used here, perhaps we can check if `ftf` is an instance of `QueryStringFunction` explicitly here and get rid of `hasFieldsInformation()`? 

If this applies to `QueryStringFunction` only, perhaps rename the method to something like `validateQueryStringFunctionPosition`? There could be a better method name, to make it more self-explained.

Reply:
For now it only applies to MatchFunction. As this is confusing, I changed it in fedefcea574f09e6260cbc233e8b936e0376420e to use `instanceof`.